File: lib/xmtp_elixir_sdk/browser_shim/opfs.ex
defmodule XmtpElixirSdk.BrowserShim.Opfs do
  @moduledoc """
  Canonical request builders for the browser-only OPFS adapter.

  The Elixir SDK owns the contract, while the browser shim owns the actual
  file-system bridge.
  """

  alias XmtpElixirSdk.BrowserShim

  defstruct enable_logging: false

  @type t :: %__MODULE__{
          enable_logging: boolean()
        }

  @spec new(keyword()) :: t()
  def new(opts \\ []) do
    %__MODULE__{
      enable_logging: Keyword.get(opts, :enable_logging, false)
    }
  end

  @spec init_request(t()) :: XmtpElixirSdk.BrowserShim.Action.Request.t()
  def init_request(%__MODULE__{enable_logging: enable_logging}) do
    BrowserShim.request("opfs.init", %{enableLogging: enable_logging})
  end

  @spec list_files_request() :: XmtpElixirSdk.BrowserShim.Action.Request.t()
  def list_files_request do
    BrowserShim.request("opfs.listFiles")
  end

  @spec file_count_request() :: XmtpElixirSdk.BrowserShim.Action.Request.t()
  def file_count_request do
    BrowserShim.request("opfs.fileCount")
  end

  @spec pool_capacity_request() :: XmtpElixirSdk.BrowserShim.Action.Request.t()
  def pool_capacity_request do
    BrowserShim.request("opfs.poolCapacity")
  end

  @spec file_exists_request(String.t()) :: XmtpElixirSdk.BrowserShim.Action.Request.t()
  def file_exists_request(path) when is_binary(path) do
    BrowserShim.request("opfs.fileExists", %{path: path})
  end

  @spec delete_file_request(String.t()) :: XmtpElixirSdk.BrowserShim.Action.Request.t()
  def delete_file_request(path) when is_binary(path) do
    BrowserShim.request("opfs.deleteFile", %{path: path})
  end

  @spec export_db_request(String.t()) :: XmtpElixirSdk.BrowserShim.Action.Request.t()
  def export_db_request(path) when is_binary(path) do
    BrowserShim.request("opfs.exportDb", %{path: path})
  end

  @spec import_db_request(String.t(), binary()) :: XmtpElixirSdk.BrowserShim.Action.Request.t()
  def import_db_request(path, data) when is_binary(path) and is_binary(data) do
    BrowserShim.request("opfs.importDb", %{path: path, data: data})
  end

  @spec clear_all_request() :: XmtpElixirSdk.BrowserShim.Action.Request.t()
  def clear_all_request do
    BrowserShim.request("opfs.clearAll")
  end
end

File: xmtp/browser_shim/src/opfs.ts
import type { OpfsAction } from "./contracts";
import { WorkerBridge } from "./WorkerBridge";

export interface OpfsTransport {
  action<A extends OpfsAction["action"]>(
    action: A,
    ...args: Extract<OpfsAction, { action: A }>["data"] extends undefined
      ? []
      : [data: Extract<OpfsAction, { action: A }>["data"]]
  ): Promise<Extract<OpfsAction, { action: A }>["result"]>;
  close(): void;
}

export class Opfs {
  #bridge: OpfsTransport;

  constructor(bridge: OpfsTransport) {
    this.#bridge = bridge;
  }

  static async create(enableLogging?: boolean) {
    const worker = new Worker(new URL("./workers/opfs", import.meta.url), {
      type: "module",
    });
    const bridge = new WorkerBridge<OpfsAction>(worker, enableLogging);
    const opfs = new Opfs(bridge as unknown as OpfsTransport);
    await opfs.init(enableLogging);
    return opfs;
  }

  async init(enableLogging?: boolean) {
    return this.#bridge.action("opfs.init", {
      enableLogging,
    });
  }

  close() {
    this.#bridge.close();
  }

  listFiles() {
    return this.#bridge.action("opfs.listFiles");
  }

  fileCount() {
    return this.#bridge.action("opfs.fileCount");
  }

  poolCapacity() {
    return this.#bridge.action("opfs.poolCapacity");
  }

  fileExists(path: string) {
    return this.#bridge.action("opfs.fileExists", { path });
  }

  deleteFile(path: string) {
    return this.#bridge.action("opfs.deleteFile", { path });
  }

  exportDb(path: string) {
    return this.#bridge.action("opfs.exportDb", { path });
  }

  importDb(path: string, data: Uint8Array) {
    return this.#bridge.action("opfs.importDb", { path, data });
  }

  clearAll() {
    return this.#bridge.action("opfs.clearAll");
  }
}
